### ISSUE TITLE
multiclass versus multiclassova in multi-class classification problems

### ISSUE THREAD

**Issue by abhishek-jha13** [Microsoft/LightGBM#1518]
For solving multi-class classification applications using LightGBM, we can choose either **multiclass** or **multiclassova** as the objective function. What is the difference between the internal implementation of these two objective functions?

Also, why in `multiclass `objective function, LightGBM should produce num_class * num_iterations trees([https://lightgbm.readthedocs.io/en/latest/Parameters.html#num_iterations](https://lightgbm.readthedocs.io/en/latest/Parameters.html#num_iterations)) as it will use softmax approach to find the probability of occurrence of a particular class?

**Comment by chivee** [comment]
Because of that lightGBM should training `N` classifier over `N` classes, and every classfier will trained by : 
     

1.  `multiclassova`: every classifier_i only learning that sample could seprated from class_i and class_rest
2.  `muticlass`: every classfier_i could learning that sample could seprated the probality over all classes. (it's not accurate acctually)

**Comment by abhishek-jha13** [comment]
Can you further explain your answer?

**Comment by guolinke** [comment]
@abhishek-jha13 
Because there is not a good multi-output tree, thus, it uses multi-trees to output multi values. 
Although there is multi-class tree, but it uses one-hot label, not the multi-value label. 
And in the gradient boosting, the training targets (residual) of multi-class is multi-value label.

**Comment by chaupmcs** [comment]
Sorry for reopen the question, but actually I  have been looking for on the Internet for all day and couldn't find any clear explanation.

I understand the way `multiclassova` works, just creates `n_class` binary classifiers, then applies the `sigmoid` function for every `n_class` raw score before returning the output. 

But when it comes to `muticlass`, I'm confused. I see it still creates `n_class` binary classifiers and uses the `softmax` at the end instead of the `sigmoid`. The main point is that the raw scores (when looking at the trees' leaves) are different from those of  `multiclassova` 's trees. It means  with `muticlass`, the Light GBM uses some variant of one-vs-rest strategy or it uses something else ?

 I don't really know how it works. Can you please explain the case `objective=muticlass`.

Thank you.

**Comment by guolinke** [comment]
there is a slight difference between multiclass and multiova.
In ova solution, the binary classifiers are independent with each other.
In softmax solution, the classifiers will affect each other, for they needs normalization the score by softmax.

**Comment by chaupmcs** [comment]
Hi @guolinke 

Thank you for the reply. I know `multiclass` use softmax to normalize the raw scores. But I dont know how it builds the tree. 

I create a model with `objective=muticlass`, and another one with `objective=muticlassova`. The two models have exactly the same parameters as well as the data input, except the `objective`. Then, I plot the first tree of each model. The raw score in the leaves of the 2 trees are not the same. 

If  the model with `objective=muticlassova` uses one-vs-rest to create n_class binary classifiers (with `log_loss` objective), what happens in case of `objective=muticlass` ?

If the difference is just the normalization, so the 2 trees should give the same raw scores in leaves.  I think I'm missing something here.

Hope you can explain a bit further.
Thank you.

**Comment by chaupmcs** [comment]
Hi @guolinke 

I've just set up the example to make it clear.  [Please go to the link here](https://github.com/chaupmcs/multiclass_vs_multiclassova/blob/master/multiclass_vs_multiclassova.ipynb). In the example, I've created 3 models namely *mutilclass_model*,  *mutilclassova_model*, and *binary_model* with exactly the same parameters and input. 

Consider the **first tree for class 1**, I see that:

**binary_model** returns the same tree compared to **mutilclassova_model** as I expected. However, the **mutilclass_model** returns different tree (different raw scores). 

Base on the result,  I think I understand how **mutilclassova_model** works. It acts just like a **binary_model**. In terms of **mutilclass_model**, I know it uses *softmax* instead of *sigmoid* function but if it's ONLY the difference, the 2 trees should still to be the same, not different like that. I really don't know what **mutilclass_model** does under the hood ? How each binary classifier in **mutilclass_model** is created ?

Can you help me out with this.

Thanks a lot.

**Comment by guolinke** [comment]
@chaupmcs 
the different raw scores can produce the same probabilities. 
For softmax, you should use all trees to get the probabilities after softmax.
BTW, there is a term in the second order gradient of softmax: https://github.com/Microsoft/LightGBM/blob/76102284d1997ef8b876cb40ae7c3e44b6ffbb66/src/objective/multiclass_objective.hpp#L86 
So, it actually cannot produce the extract same results with ova.

**Comment by chaupmcs** [comment]
Thanks. It helped me a lot.
